### DISCUSSION TITLE
systemd timers for rootless containers doesn't work

### DISCUSSION THREAD

**Discussion by ghost** [containers/podman#20364] [General]
hello community, 

1)  I have created a new linux user without root privileges 
```
sudo useradd -m username uptime
sudo passwd uptime
```

2) I have created a new quadlet file
```
mkdir -p ~/.config/containers/systemd/
nano ~/.config/containers/systemd/uptime.container
```
Content
```
[Unit]
Description=uptime-kuma
After=local-fs.target

[Container]
ContainerName=uptime-kuma
Image=docker.io/louislam/uptime-kuma
AutoUpdate=registry
PublishPort=3001:3001
Volume=uptime-kuma:/app/data

[Install]
WantedBy=default.target
```

3) I have reloaded and started the service
```
systemctl --user daemon-reload
systemctl --user start uptime
```

Now i am trying to auto-update the container daily at 02:00:00, so i am editing the podman-timer with the following command
`systemctl --user edit podman-auto-update.timer` and this content:

```
[Unit]
Description=Podman auto-update timer

[Timer]
OnCalendar=*-*-* 02:00:00
Persistent=true

[Install]
WantedBy=timers.target
```

Whole File:
```
### Editing /home/uptime/.config/systemd/user/podman-auto-update.timer.d/override.conf
### Anything between here and the comment below will become the contents of the drop-in file

[Unit]
Description=Podman auto-update timer

[Timer]
OnCalendar=*-*-* 02:00:00
Persistent=true

[Install]
WantedBy=timers.target

### Edits below this comment will be discarded


### /usr/lib/systemd/user/podman-auto-update.timer
# [Unit]
# Description=Podman auto-update timer
#
# [Timer]
# OnCalendar=daily
# RandomizedDelaySec=900
# Persistent=true
#
# [Install]
# WantedBy=timers.target
```

But when i now reload and watch the status of my timer it still shows the old standard timer at midnight every day.

```
systemctl --user daemon-reload
systemctl --user status podman-auto-update.{service,timer}
```
Output:
```
○ podman-auto-update.service - Podman auto-update service
     Loaded: loaded (/usr/lib/systemd/user/podman-auto-update.service; disabled; preset: enabled)
     Active: inactive (dead)
TriggeredBy: ● podman-auto-update.timer
       Docs: man:podman-auto-update(1)

Oct 15 15:37:43 arch systemd[282]: Starting Podman auto-update service...
Oct 15 15:37:43 arch systemd[282]: Finished Podman auto-update service.
Oct 15 15:38:07 arch systemd[282]: Starting Podman auto-update service...
Oct 15 15:38:07 arch systemd[282]: Finished Podman auto-update service.

● podman-auto-update.timer - Podman auto-update timer
     Loaded: loaded (/usr/lib/systemd/user/podman-auto-update.timer; enabled; preset: enabled)
    Drop-In: /home/uptime/.config/systemd/user/podman-auto-update.timer.d
             └─override.conf
     Active: active (waiting) since Sun 2023-10-15 16:21:02 CEST; 3min 22s ago
    Trigger: Mon 2023-10-16 00:01:49 CEST; 7h left
   Triggers: ● podman-auto-update.service

```
What am I missing?

Regards

**Comment by vrothberg** [comment]
Thanks for reaching out!

Try running `systemctl --user list-timers` and check whether the output matches what you expect.

**Comment by ghost** [comment]
Thank you for your fast response @vrothberg !

That is my output:
```
NEXT                         LEFT LAST                         PASSED UNIT                     ACTIVATES               
  
Tue 2023-10-17 00:04:39 CEST  12h Mon 2023-10-16 11:52:18 CEST 6s ago podman-auto-update.timer podman-auto-update.service
```
So, sadly that is not what i expect.

**Reply by vrothberg** [reply]
It works on my machine.  Maybe there's something off on yours but that exceeds my systemd skills.

**Comment by Luap99** [comment]
Note that systemd allows you to set more than one `OnCalendar` option, see https://www.freedesktop.org/software/systemd/man/systemd.timer.html#OnCalendar=
And if you use systemd edit to create these drop in files they are just that a drop in. They do not fully replace the original unit.

Thus your example now has two OnCalendar intervals specified at midnight and 2 am.
So the proper drop in should look like this:
```
[Timer]
OnCalendar=
OnCalendar=*-*-* 02:00:00
```
The empty OnCalendar is used to unset all prior OnCalendar options.

**Reply by ghost** [reply]
That is the solution! Thank You.
In my opinion that should be added to the podman documentation. Although I am aware that it is not podman thing.

Let me ask you another question:
Do I have to execute the following two commands if I have already activated the timer?
```
systemctl --user start --now podman-auto-update.service
systemctl --user enable --now podman-auto-update.service
```
What is the exact difference between the .service and .timer? What happens if i only enable / start the timer resp. the service?

**Reply by Luap99** [reply]
Feel free to open a docs PR if you want to document it for others.

>  Do I have to execute the following two commands if I have already activated the timer?

No you should not enable the service, this means podman auto-update will be run on boot which does not make a lot of sense unless this is what you want. If you start the service you basically execute podman auto-update right away. 

> What is the exact difference between the .service and .timer? What happens if i only enable / start the timer resp. the service?

A service unit describes how to run a program. The timer describes when to run said service.
You need to start the timer unit in order for the timer to be active and then later trigger the service start. The enable part is necessary in order to have it automatically started on boot (or technically only according to the [Install] section).

In any way this is not podman specific so I recommend you refer to systemd docs/blogs for more information.
